### PR TITLE
Explicit device tensors

### PR DESCRIPTION
**Pull request by kpot**
### Checklist

- [x] Confirmed that `run-checks all` script has been executed.
- [X] Made sure the book is up to date with changes in this PR.

### Related Issues/PRs

Covers renaming and related refactoring as it was agreed and described in #518

### Changes

Allows for explicit creation of tensors on specific devices. All public interfaces (of layers, models, etc.) have been refactored accordingly, and now allow to specify a `Backend::Device` during initialization. Onnx macroses support generation of device-aware models too.

### Testing

All the tests I could find and run are passing, including

- `cargo test`
- `./run-checks.sh std`
- `./run-checks.sh no-std`
- `./run-checks.sh typos`
- `./run-checks.sh examples`

### CODE REVIEW

File: burn-core/src/optim/grads.rs [hunk 1]
```diff
@@ -115,11 +116,11 @@ mod tests {
...
-    fn layer() -> Linear<TestAutodiffBackend> {
-        LinearConfig::new(20, 20).with_bias(true).init()
+    fn layer<B: Backend>(device: &B::Device) -> Linear<B> {
+        LinearConfig::new(20, 20).with_bias(true).init(device)
     }
 
-    fn random_tensor() -> Tensor<TestAutodiffBackend, 2> {
-        Tensor::<TestAutodiffBackend, 2>::random([2, 20], Distribution::Default)
+    fn random_tensor<B: Backend>(device: &B::Device) -> Tensor<B, 2> {
+        Tensor::<B, 2>::random([2, 20], Distribution::Default, &device)
```
Comment (nathanielsimard) [inline burn-core/src/optim/grads.rs:124]: I think you can remove the extra `&` before `device` here.

**Review by nathanielsimard — APPROVED** [review]
Thanks a lot for this work! I think it's going to be much less likely to have different device errors.

There is some place in `burn-import` where we use `from_data_default` and not `from_data`; not sure what should be the proper usage there, @antimora.

Approved from my side, but I would appreciate one more set of eyes on this, @antimora.

**Review by antimora — CHANGES_REQUESTED** [review]
Oooh, that's a big change in the APIs.

Although I'm reviewing this a bit late, I understand and respect the design decisions made by the team. However, to mitigate potential disruption for our existing users, I'd like to request the following adjustments:

1. All `.init` module methods (in core; the examples can remain as is) should be accompanied by corresponding `.init_default` methods. These should call `.init` with the default device. This approach will allow for quick and easy refactoring of existing model code, essentially requiring just the addition of a `_default` suffix to the method calls. This suggestion aligns well with the changes you've made to other Tensor methods, which are undergoing similar refactoring.

2. In a similar fashion, the model generated from ONNX in the `burn-import/src/burn/graph.rs` file should include a `new_default()` method. This method would call `new` with the default device. If you could adjust the tests, it'd be great too but this more cleanliness and less verbosity.

**Comment by kpot** [comment]
Hi!
@antimora @nathanielsimard I generally agree that something like `init_default` can be useful, but `new_default`... It is not very idiomatic to Rust naming conventions. In Rust, whenever you want to construct something with default settings, you simply implement `Default` trait, and create an instance by calling the `default()` method.

So, I propose to alter the naming. Again. Let's replace the vague `_default` suffix by `_autodev`. It hints at its purpose: automatic selection of device. And it has the same number of characters as `_default`, thus making it possible to do the refactoring more or less using a simple "find and replace" tool, without breaking code formatting.

Another decent names could be be 

- `_autoloc` (automatic location, to me sounds as neutral as `_autodev`), 
- `_adevice` (any device, hints at randomness)
- `_bestdev` (the best device we can choose, implies the choice is not random)
- `_somedev` (hits at an arbitrary choice of the device too)

but you can come up with something even better. What do you think?

**Comment by antimora** [comment]
Sounds good. My top preferences:

- '_defaultdev'
- '_devauto'
- '_devdefault'
- '_autodev'
- '_devault'

**Comment by kpot** [comment]
@antimora Hi! All the changes you requested were implemented. Specifically:

1. I chose `_devauto` suffix as a replacement for the `_default`. 
2. All modules that have their own weights now also have a `init_devauto()` method.
3. ONNX models now include `new_devauto()` method as well
4. All introduced `*_devauto` methods have at least one test covering them.

All the tests and checks I know of are passing.

**Comment by nathanielsimard** [comment]
@kpot, I don't think the new naming is correct. First, "automatic" is associated with something intelligent and not error-prone, so "auto device" isn't the same as "default device." Second, I don't think using `dev` for "device" is actually a good idea; it's normally used for development, and it will be very confusing!

However, I agree that using `_default` isn't the best naming convention, so let me propose something a bit radical: remove all functions with the default device! It will make the API smaller, and someone can always use `Default::default()` explicitly where it applies. I think the only place where it makes sense to use the default device is in tests, not in production code where it can cause device errors.

Enforcing users to specify the device where it's required is not a bad idea, even if it's more verbose.

The only functions where the default device could be used is when implementing the `From` trait. So someone could still use the following:

```rust
let tensor = Tensor::from([5.0, 10.0]);
```

I believe this could be used for most tests. What is your opinion, @kpot, @antimora, @louisfd?

**Comment by kpot** [comment]
@nathanielsimard Well, this was my [original proposal](https://github.com/tracel-ai/burn/issues/518#issuecomment-1806015115) too, to get rid of all the twin `*_device` methods and make them a new norm. Any constructor that internally creates some device-bound parameters must be given a device explicitly. Any half-measures like `_devauto` will inevitably lead to more bugs. So yeah, I support your suggestion. This will require a lot of refactoring within tests (yay, more refactoring!) but it's worth it.

@antimora Do you approve?

**Comment by antimora** [comment]
> @nathanielsimard Well, this was my [original proposal](https://github.com/tracel-ai/burn/issues/518#issuecomment-1806015115) too, to get rid of all the twin `*_device` methods and make them a new norm. Any constructor that internally creates some device-bound parameters must be given a device explicitly. Any half-measures like `_devauto` will inevitably lead to more bugs. So yeah, I support your suggestion. This will require a lot of refactoring within tests (yay, more refactoring!) but it's worth it.
> 
> @antimora Do you approve?

Yes, I guess in the end it would be the right solution. So I approve it. 

I appreciate your willingness to go through big changes and stand by the right decision. I myself had to scrap a big chunk of code during a Burn PR review a few times.

**Comment by nathanielsimard** [comment]
@kpot I think it would be easier to merge the PR right now as is and refactor the tests in other PRs. We only have to remove all occurrences of `autodev` in the codebase before deleting the functions. This way, it minimizes git conflicts as well as making the upcoming PR reviewable! What do you think, @antimora?

**Review by antimora — APPROVED** [review]

**Comment by antimora** [comment]
> @kpot I think it would be easier to merge the PR right now as is and refactor the tests in other PRs. We only have to remove all occurrences of `autodev` in the codebase before deleting the functions. This way, it minimizes git conflicts as well as making the upcoming PR reviewable! What do you think, @antimora?

Yep. Lets go ahead and merge it!

**Pull request merged**